// @toolhouseai/cli/lib/utils/telemetry.js
const fetch = require('node-fetch');

const TELEMETRY_URL = `${process.env.TOOLHOUSE_API_URL}/v1/cli-telemetry`;

let sessionId = null;
let sessionEvents = [];
let sessionStarted = false;

function generateSessionId() {
  // Simple random string for session ID
  return (
    Date.now().toString(36) + '-' + Math.random().toString(36).substring(2, 10)
  );
}

function startTelemetrySession() {
  if (!sessionStarted) {
    sessionId = generateSessionId();
    sessionEvents = [];
    sessionStarted = true;
  }
}

function recordTelemetryEvent(event) {
  if (!sessionStarted) startTelemetrySession();
  sessionEvents.push({
    ...event,
    timestamp: new Date().toISOString(),
  });
}

async function endTelemetrySession() {
  if (process.env.TELEMETRY_DISABLED_ENV === '1' || !sessionStarted) return;
  try {
    await fetch(TELEMETRY_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        session_id: sessionId,
        events: sessionEvents,
      }),
    });
  } catch (err) {
    // Fail silently, do not block CLI
  } finally {
    sessionStarted = false;
    sessionId = null;
    sessionEvents = [];
  }
}

module.exports = {
  startTelemetrySession,
  recordTelemetryEvent,
  endTelemetrySession,
};
