// @toolhouseai/cli/lib/commands/new.js
const fs = require("fs");
const path = require("path");
const chalk = require("chalk");
const { input } = require("@inquirer/prompts");
const { createSpinner } = require("../utils/spinner");
const {
  formatSuccessBox,
  formatErrorMessage,
  formatErrorBox,
} = require("../utils/display");
const {
  generateRandomName,
  createSampleYamlContent,
} = require("../utils/generator");

function validateAgentName(name) {
  return /^[a-zA-Z0-9-]+$/.test(name);
}

async function promptForName(initialName) {
  console.log(
    `${chalk.dim("›")} Generated unique name: ${chalk.cyan.bold(
      initialName
    )}\n` +
      `${chalk.dim("│")} ${chalk.dim(
        "Only letters, numbers, and dashes allowed"
      )}\n`
  );

  const customName = await input({
    message: "Enter a custom name or press Enter to accept",
    default: initialName,
    validate: (value) => {
      const name = value.trim().replace(/\.yaml$/, "");

      if (!validateAgentName(name)) {
        return formatErrorMessage(
          "Invalid name. Only letters, numbers, and dashes allowed"
        );
      }

      if (fs.existsSync(`${name}.yaml`)) {
        return formatErrorMessage(
          `Duplicate agent file. An agent file named '${name}' already exists in this directory.`
        );
      }

      return true;
    },
  });

  return customName.trim().replace(/\.yaml$/, "");
}

function newCommand(program) {
  program
    .command("new [agent_name]")
    // TODO: support agent in directory creation so you can create a new agent in a specific directory
    .description("Create a new agent file")
    .action(async (agentName) => {
      try {
        let fileName;
        // If agentName is not provided, generate a random name
        if (!agentName) {
          fileName = generateRandomName();
          while (fs.existsSync(`${fileName}.yaml`)) {
            fileName = generateRandomName();
          }

          fileName = await promptForName(fileName);
        } else {
          fileName = agentName.replace(/\.yaml$/, "");

          if (!validateAgentName(fileName)) {
            console.error(
              formatErrorMessage(
                "Invalid name. Use only letters, numbers, and dashes."
              )
            );
            process.exit(1);
          }

          if (fs.existsSync(`${fileName}.yaml`)) {
            console.error(
              formatErrorMessage(
                `Duplicate agent file. An agent file named '${fileName}' already exists in this directory.`
              )
            );
            process.exit(1);
          }
        }

        const spinner = createSpinner(chalk.bold("Creating your new agent"));
        spinner.start();

        const filePath = path.join(process.cwd(), `${fileName}.yaml`);
        fs.writeFileSync(filePath, createSampleYamlContent());

        spinner.succeed(
          chalk.bold(
            `Agent file created: ${chalk.green.bold(fileName) + ".yaml"}`
          )
        );

        console.log(
          formatSuccessBox(
            "Agent File Created",
            {
              Name: fileName,
              "Agent file": filePath,
            },
            {
              "Run your agent": `th run ${fileName}.yaml`,
              "Deploy your agent": `th deploy ${fileName}.yaml`,
            }
          )
        );
      } catch (error) {
        formatErrorMessage(
          `An error occurred while creating the agent file: ${error.message}`
        );
        process.exit(1);
      }
    });
}

module.exports = newCommand;
